dango/dex/src/core/router.rs

```diff
@@ -0,0 +1,80 @@
+use {
+    crate::{PAIRS, PassiveLiquidityPool, RESERVES},
+    dango_types::dex::{PairId, SwapExactAmountIn, SwapExactAmountOut},
+    grug::{Coin, CoinPair, EventBuilder, Inner, NonZero, Storage, UniqueVec},
+    std::collections::HashMap,
+};
+
+pub fn swap_exact_amount_in(
+    storage: &dyn Storage,
+    route: UniqueVec<PairId>,
+    input: Coin,
+) -> anyhow::Result<(HashMap<PairId, CoinPair>, Coin, EventBuilder)> {
+    let mut reserves = HashMap::new();
+    let mut output = input;
+
+    let mut events = EventBuilder::new();
+
+    for pair in route.into_iter() {
+        // Load the pair's parameters.
+        let params = PAIRS.load(storage, (&pair.base_denom, &pair.quote_denom))?;
+
+        // Load the pool's reserves.
+        let mut reserve = RESERVES.load(storage, (&pair.base_denom, &pair.quote_denom))?;
+
+        // Clone the input for the event.
+        let input = output.clone();
+
+        // Perform the swap.
+        // The output of the previous step is the input of this step.
+        (reserve, output) = params.swap_exact_amount_in(reserve, output)?;
+
+        // Save the updated pool reserve.
+        reserves.insert(pair.clone(), reserve);
+
+        // Add events
+        events.push(SwapExactAmountIn {
+            pair,
+            input,
+            output: output.clone(),
+        })?;
```

Comment:
I feel we should have only one event regardless of how many step there are in the route. 

The purpose of events if for notifications on the frontend.

If I swap A -> B -> C, I want to only receive one notification "you swapped A for C", not two notifications "you swapped A for B" and "you swapped B for C",